### PR TITLE
Point restructure to allow conversion between LAS point formats

### PR DESCRIPTION
**Pull request by leo-stan**

### CODE REVIEW

**Comment by CCInc** [comment]
A concern I have with `point14` and `point10` as the underlying format is that performing a `ToPointFormat` on a list of them requires copying each point to new memory, whereas if we had a single `point` object we could do it without copying them. Since most of the differences between Point14 and Point10 are just enlarging/shrinking the size of some fields, we could reduce the problem space from deleting/recreating all point objects in memory to changing some values that would cause an overflow.

For example, if going from point10 -> point14, and we had one "point" object, we wouldn't need to do any conversion (we can store the scanangle as the point14 format for all point types). If going from point14 -> point10, we can just iterate through fields like classification and return angle and either reset them to 0 or cap them at the maximum. But this operation can happen in-place.

**Comment by CCInc** [comment]
tldr I think we should change our implementation to be basically akin to laszip's LasPoint object (we could just copy most of it tbh, it does almost exactly what we need 😄 ). He already handles point14 vs point10 and the rgb/gpstime attribute correctly.

Then, all we'll have to do is add some getter/setter logic around get_extended_classification and get_classification based on `extended_point_type` and we should be good. Let me know your thoughts.

https://github.com/LAStools/LAStools/blob/master/LASzip/src/laspoint.hpp

**Comment by leo-stan** [comment]
I question remaining is about the behaviour when going from point14 to point10, what should we do if classification>31, returns>7, scan angle >90, etc.

**Review by CCInc — COMMENTED** [review]
Great work! I added few styling comments.

Can you add some test cases for going to larger/smaller point formats? Such as from 0 to 3 and back. So we can make sure that the fields that we add are properly zeroed out. 

Do we initialize the fields to a default anywhere when we create the point object? If not we should initialize them all to 0. Then, when we convert, we can zero out all the bitfields that we are changing before we compute them so that there won't be any stale data. 

We could also add some tests for pack/unpack as well - initializing a new point of each format, then packing it and verifying the byte arrays are zeroed properly, making sure there's no random byte values happening etc.

**Comment by leo-stan** [comment]
I currently have a default constructor for Point, initializing everything to 0.

File: test/point_test.cpp [hunk 1]
```diff
@@ -315,30 +313,99 @@ TEST_CASE("Point tests", "[Point]")
...
+        REQUIRE(point.Withheld() == true);
+        REQUIRE(point.Overlap() == false);
+        REQUIRE(point.ExtendedScanAngle() == 7500);
+        REQUIRE(point.ScannerChannel() == 0);
+
+        point.ToPointFormat(0);
+
+        REQUIRE(point.ReturnNumber() == 5);
+        REQUIRE(point.NumberOfReturns() == 6);
+        REQUIRE(point.ScanDirectionFlag() == false);
```
Comment (CCInc) [inline test/point_test.cpp:379]: can we add some REQUIRE_THROWS tests for the converted points, such as point.overlap()?

**Review by CCInc — COMMENTED** [review]
cool! must have missed that.

**Comment by leo-stan** [comment]
I don't actually 0 out the values when going to a more restrictive point type, since you can't access them anyways. For instance if you go from 3 to 0 to 3, you will still have your GPS value, it won't be zeroed out when going to format 0. Instead when you're under format 0 you just can't call the getter for GPS.

Should I change this behavior?

**Comment by CCInc** [comment]
It should be okay for gps/rgb, as long as it's not random initialized. Lmk when you're ready for review

**Comment by CCInc** [comment]
I wonder if we should hide/rename some of the bit fields to make them less accessible to the user, so they don't get confused between say returnnumber and extendedreturns, what do you think? We could either not expose them at all or maybe change the naming schema for them, unless you can think of another way?

**Comment by leo-stan** [comment]
The bitfield access is pretty handy to avoid making multiple operations when reassigning the whole bitfield. Could change the name I guess and call it BitField.

File: src/las/point.cpp [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        has_rgb_ = false;
+    if (FormatHasNir(point_format_id))
+    {
+        nir_ = internal::unpack<uint16_t>(in_stream);
+        has_nir_ = true;
+    }
+    else
+        has_nir_ = false;
+
+    point_byte_size_ = BaseByteSize(point_format_id);
```
Comment (CCInc) [inline src/las/point.cpp:61]: point_byte_size_ = point_record_length

File: src/las/point.cpp [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        // Scan angle
+        extended_scan_angle_ = static_cast<int16_t>(float(ScanAngleRank()) / 0.006f);
+    }
+
+    // Update flags
+    extended_point_type_ = point_format_id > 5;
+    has_gps_time_ = FormatHasGpsTime(point_format_id);
+    has_rgb_ = FormatHasRgb(point_format_id);
+    has_nir_ = FormatHasNir(point_format_id);
+    point_byte_size_ = BaseByteSize(point_format_id);
```
Comment (CCInc) [inline src/las/point.cpp:176]: `+ len(extended bytes)`

File: include/copc-lib/las/point.hpp [hunk 4]
```diff
@@ -1,597 +1,590 @@
...
 
-    uint16_t PointSourceId() const { return point_10_ ? point_10_->PointSourceId() : point_14_->PointSourceId(); }
-    void PointSourceId(const uint16_t &point_source_id)
+    uint8_t UserData() const { return user_data_; }
+    void UserData(const uint8_t &user_data) { user_data_ = user_data; }
+
+    uint16_t PointSourceId() const { return point_source_id_; }
+    void PointSourceId(const uint16_t &point_source_id) { point_source_id_ = point_source_id; }
+
+    const uint16_t *Rgb() const
```
Comment (CCInc) [inline include/copc-lib/las/point.hpp:294]: We can change Rgb to RGB everywhere it's capitalized (allowed by style guide for abbreviations)

File: include/copc-lib/las/point.hpp [hunk 5]
```diff
@@ -1,597 +1,590 @@
...
+        {
+            rgb_[0] = r;
+            rgb_[1] = g;
+            rgb_[2] = b;
+        }
         else
             throw std::runtime_error("This point format does not have RGB.");
     }
 
     uint16_t R() const
```
Comment (CCInc) [inline include/copc-lib/las/point.hpp:316]: Maybe we can name these Red, Green, Blue?

File: include/copc-lib/las/point.hpp (hunk 4 above)
Comment (CCInc) [reply include/copc-lib/las/point.hpp:294]: Should we allow the user to change this array? Or do you want to make them use the setter?

File: include/copc-lib/las/point.hpp [hunk 6]
```diff
@@ -1,597 +1,590 @@
...
+        case 2:
+            return false;
+        case 3:
+        case 4:
+        case 5:
+        case 6:
+        case 7:
+        case 8:
+        case 9:
+        case 10:
```
Comment (CCInc) [inline include/copc-lib/las/point.hpp:420]: Let's just error out on all the waveform formats I think

File: include/copc-lib/las/point.hpp [hunk 7]
```diff
@@ -1,597 +1,590 @@
...
+            return true;
+        default:
+            throw std::runtime_error("Point format must be 0-10");
```
Comment (CCInc) [inline include/copc-lib/las/point.hpp:423]: In the future we should look at adding our own exception class 😄

File: include/copc-lib/las/point.hpp [hunk 8]
```diff
@@ -1,597 +1,590 @@
...
     {
-        if (point_10_)
-            point_10_->Classification(classification);
+        if (extended_point_type_)
+            extended_classification_ = classification;
         else
-            point_14_->Classification(classification);
+        {
+            if (classification > 31)
+                throw std::runtime_error("Classification for Point10 must be <= 31");
```
Comment (CCInc) [inline include/copc-lib/las/point.hpp:225]: Add a note to the exception that if you want to change the point10 raw classification (or want to override this check), use ClassificationBitField

File: include/copc-lib/las/point.hpp [hunk 9]
```diff
@@ -1,597 +1,590 @@
...
 {
   public:
-    Point(const uint8_t &point_format_id)
-    {
-        point_10_ = nullptr;
-        point_14_ = nullptr;
-        rgb_ = nullptr;
-        gps_time_ = nullptr;
-        nir_ = nullptr;
+    Point() = default;
```
Comment (CCInc) [inline include/copc-lib/las/point.hpp:13]: let's make this constructor private and add a public constructor that takes the point format and point size as args

File: include/copc-lib/las/point.hpp [hunk 10]
```diff
@@ -1,597 +1,590 @@
...
-        case 10:
-            point_14_ = new internal::Point14();
-            rgb_ = new internal::Rgb();
-            gps_time_ = new internal::GpsTime();
-            nir_ = new internal::Nir();
-            break;
-        default:
-            throw std::runtime_error("Point format must be 0-10");
-        }
+    void Unpack(std::istream &in_stream, const uint16_t &point_format_id, const uint16_t &point_record_length);
```
Comment (CCInc) [inline include/copc-lib/las/point.hpp:15]: I think this should be a static function that returns a Point

**Review by CCInc — CHANGES_REQUESTED** [review]

File: include/copc-lib/las/point.hpp (hunk 8 above)
Comment (leo-stan) [reply include/copc-lib/las/point.hpp:225]: It can't be over 31 though since there are only 5 bits available in the bitfield?

File: src/las/point.cpp (hunk 3 above)
Comment (leo-stan) [reply src/las/point.cpp]: Should the point_record_length_ include the size of extra bytes?

File: include/copc-lib/las/point.hpp (hunk 10 above)
Comment (leo-stan) [reply include/copc-lib/las/point.hpp:15]: So separate from the Point class?

File: include/copc-lib/las/point.hpp (hunk 9 above)
Comment (leo-stan) [reply include/copc-lib/las/point.hpp:13]: Currently I get that information while unpacking, you're suggesting we would provide that info at construction and then use it when unpacking?

File: src/las/point.cpp (hunk 3 above)
Comment (leo-stan) [reply src/las/point.cpp]: I currently just have it as the minimum point size, and the number of extra bytes can be accessed by calling ExtraBytes().size()

Comment (leo-stan) [reply src/las/point.cpp]: I'm also planning to change from vector to char* for extra bytes, I think the vectors may take more memory when we don't really need their features.

Comment (leo-stan) [reply src/las/point.cpp]: In which case I can have point_record_length_ as the length of the point data, and either a variable storing the number of extra bytes, or a variable storing the total byte size of point+EBs. What do you think?

File: include/copc-lib/las/point.hpp (hunk 10 above)
Comment (CCInc) [reply include/copc-lib/las/point.hpp:15]: It can be within the point class. But it should be a static function that returns a point object. See here: https://github.com/RockRobotic/copc-lib/blob/writer/include/copc-lib/hierarchy/entry.hpp

File: include/copc-lib/las/point.hpp (hunk 8 above)
Comment (CCInc) [reply include/copc-lib/las/point.hpp:225]: Yes but technically you can set all 8 bits. Some viewers still show that as a "classification", such as the ones that don't support flags. So sometimes you want to treat it as an 8 bit field. It might be good just to make a note of it.

File: src/las/point.cpp (hunk 3 above)
Comment (CCInc) [reply src/las/point.cpp]: I think point_record_length_  should include the extra byte size, to make it consistent with the LAS spec. What lazperf does internally is stores the point_record_length inluding the extra bytes, then on functions where it needs to know the original point size, you pass in the point_record_id and the number of extra bytes (computed by point_record_length - BaseSize(point_record_id). 

I think the vectors have pretty low overhead and are "safer" than char* objects. I think it's good to err on the side of c++ oriented code, trying to avoid C features. But let me know what you think.

File: include/copc-lib/las/point.hpp (hunk 9 above)
Comment (CCInc) [reply include/copc-lib/las/point.hpp:13]: Yeah, something like that. So the user can create points of a certain point type manually. Something like this:

```
static Point Unpack(in_stream, point_id, point_length)
{
	Point point(point_id, point_length);
	x_ = ...
}
```

To enable Unpack to be able to set the private fields of the point, you can mark it as a friend function (see here for static functions https://stackoverflow.com/a/19845672).

File: include/copc-lib/las/point.hpp (hunk 8 above)
Comment (leo-stan) [reply include/copc-lib/las/point.hpp:225]: ok

File: src/las/point.cpp (hunk 3 above)
Comment (leo-stan) [reply src/las/point.cpp]: Cool, I don't really like char* either haha. Wasn't sure how much overhead vectors added. Happy to keep 'em!

**Review by CCInc — APPROVED** [review]
